src/lib.rs

```diff
@@ -570,14 +579,86 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for Evaluator<'tcx> {
     fn stack_push(
         ecx: &mut InterpretCx<'mir, 'tcx, Self>,
     ) -> InterpResult<'tcx, stacked_borrows::CallId> {
-        Ok(ecx.memory().extra.borrow_mut().new_call())
+        Ok(ecx.memory().extra.stacked.borrow_mut().new_call())
     }
 
     #[inline(always)]
     fn stack_pop(
         ecx: &mut InterpretCx<'mir, 'tcx, Self>,
         extra: stacked_borrows::CallId,
     ) -> InterpResult<'tcx> {
-        Ok(ecx.memory().extra.borrow_mut().end_call(extra))
+        Ok(ecx.memory().extra.stacked.borrow_mut().end_call(extra))
+    }
+
+    fn int_to_ptr(
+        int: u64,
+        memory: &Memory<'mir, 'tcx, Self>,
+    ) -> InterpResult<'tcx, Pointer<Self::PointerTag>> {
+        if int == 0 {
+            return err!(InvalidNullPointerUsage);
+        }
+        
+        if memory.extra.seed.is_none() {
+            return err!(ReadBytesAsPointer);
+        }
+
+        let extra = memory.extra.intptrcast.borrow();
+        
+        match extra.vec.binary_search_by_key(&int, |(int, _)| *int) {
+            Ok(pos) => {
+                let (_, alloc_id) = extra.vec[pos];
+                Ok(Pointer::new_with_tag(alloc_id, Size::from_bytes(0), Tag::Untagged))
+            }
+            Err(pos) => {
+                if pos > 0 {
+                    let (glb, alloc_id) = extra.vec[pos - 1];
+                    let offset = int - glb;
+                    if offset <= memory.get(alloc_id)?.bytes.len() as u64 {
+                        Ok(Pointer::new_with_tag(alloc_id, Size::from_bytes(offset), Tag::Untagged))
+                    } else {
+                        return err!(DanglingPointerDeref);
+                    }
+                } else {
+                    return err!(DanglingPointerDeref);
+                }
+            }
+        }
+    }
+ 
+    fn ptr_to_int(
+        ptr: Pointer<Self::PointerTag>,
+        memory: &Memory<'mir, 'tcx, Self>,
+    ) -> InterpResult<'tcx, u64> {
+        if memory.extra.seed.is_none() {
+            return err!(ReadPointerAsBytes);
+        }
+
+        let mut extra = memory.extra.intptrcast.borrow_mut();
+
+        let alloc = memory.get(ptr.alloc_id)?;
+
+        let mut base_addr = alloc.extra.base_addr.borrow_mut();
+        
+        let addr = match *base_addr { 
+            Some(addr) => addr,
+            None => {
+                let addr = extra.addr;
+                extra.addr += alloc.bytes.len() as u64;
+
+                *base_addr = Some(addr);
+
+                let elem = (addr, ptr.alloc_id);
+
+                if let Err(pos) = extra.vec.binary_search(&elem) {
```

Comment:
This can't happen. You can just replace this entire `if` statement with `extra.vec.push((addr, ptr.alloc_id));` If you want to keep it as a sanity check, do `.unwrap_err()` on the binary search result and then assert that `pos == extra.vec.len()` since we should always be inserting at the end.